### PR TITLE
caffe2 ONNX import  Upsample(mode=bilinear)-> UpsampleBilinear

### PR DESCRIPTION
**Pull request by 0wu**
There are two types of ONNX [Upsample](https://github.com/onnx/onnx/blob/master/docs/Operators.md#Upsample
) (mode=nearest) and Upsample(mode=linear)
Caffe2 used to import all ONNX Upsample to ResizeNearest
This PR adds support for:
* ONNX import
  Upsample(mode=linear) -> UpsampleBilinear,
  Upsample(mode=nearest) -> ResizeNearest
* ONNX export of the above two caffe2 operators
* a unit-test case is added to verify bidirectional translation.

### CODE REVIEW

**Comment by 0wu** [comment]
@ezyang checks are passing now, please let me know if there's anything needs to be done.

File: caffe2/python/onnx/tests/c2_ref_test.py [hunk 1]
```diff
@@ -102,6 +102,38 @@ def sigmoid(x):
...
+            init_net=None,
+            predict_net=predict_net,
+            inputs=[X])
+
+        onnx_model = c2_onnx.caffe2_net_to_onnx_model(
+            predict_net=predict_net,
+            value_info={
+                'X': (onnx.mapping.NP_TYPE_TO_TENSOR_TYPE[X.dtype], X.shape)
+            })
+        print(onnx_model)
```
Comment (ezyang) [inline caffe2/python/onnx/tests/c2_ref_test.py]: I don't think you want to actually print this as part of the test

**Comment by ezyang** [comment]
@houseroad, maybe you could review this?

**Comment by houseroad** [comment]
@ezyang yeah, will do it.

File: caffe2/onnx/backend.cc [hunk 2]
```diff
@@ -189,6 +189,17 @@ float OnnxAttributes::get(const std::string& key) const {
   return value;
 }
 
+template <>
+std::string OnnxAttributes::get(const std::string& key) const {
```
Comment (houseroad) [inline caffe2/onnx/backend.cc]: overloading with return type is dangerous. can we change the existing mechanism and use i (int), f (float), s (string), etc?

File: caffe2/onnx/backend.cc [hunk 3]
```diff
@@ -1256,10 +1268,19 @@ Caffe2Ops Caffe2Backend::CreateUpsample(
         {arg_starts, arg_ends});
 
     // Upsample
+    std::string c2_op_name;
+    if (mode == "linear") {
+        c2_op_name = "UpsampleBilinear";
+    } else if (mode == "nearest") {
+        c2_op_name = "ResizeNearest";
+    } else {
+        VLOG(2) << "Operator " << node.name() << " contains unsupported mode " << mode << std::endl;
```
Comment (houseroad) [inline caffe2/onnx/backend.cc]: CAFFE_THROW here? Since it cannot be handled.

**Review by houseroad — COMMENTED** [review]
Thanks for the contribution. Could you also add some test cases for 'bilinear' mode here: https://github.com/pytorch/pytorch/blob/master/test/onnx/test_pytorch_onnx_caffe2.py#L764?

File: caffe2/onnx/backend.cc (hunk 2 above)
Comment (sampepose) [reply caffe2/onnx/backend.cc]: @houseroad can you elaborate on what changes you want to see here? Overloading with return type is the status quo for` OnnxAttributes::get()`, perhaps refactoring this mechanism should go in a new PR?

**Comment by garymm** [comment]
Given the age of the PR and the state of caffe2 development, I suggest we close this.
@0wu would you close it? If not, I guess try to get someone to review it.

**Pull request closed without merging**